src/Functions/FunctionSnowflake.h

```diff
@@ -5,6 +5,7 @@
 #include <Functions/FunctionHelpers.h>
 #include <DataTypes/DataTypeDateTime64.h>
 #include <DataTypes/DataTypesNumber.h>
+#include <Columns/ColumnConst.h>
```

Comment:
About LowCardinality and Sparse columns: The function framework unpacks them automatically before calling the changed code (`IFunction::useDefaultImplementationForLowCardinalityColumns`, `useDefaultImplementationForSparseColumns`).

About the other functions in this cpp: yes, there is an issue. Fixing it.

Reply:
Update: Functions `dateTime*ToSnowflake()` set `useDefaultImplementationForConstants() == true` which means that const columns are automatically unpacked. So we don't need to fix anything else for them.

Reply:
SnowflakeToDateTime*() functions have `bool useDefaultImplementationForConstants() const override { return true; }` as well. Is it not effective for some reason?

Reply:
I thought the same first. The explanation is a bit complicated.

Any SQL function can have const and non-const arguments (ignore sparse and low-cardinality because they are in 99.9% of all cases automatically uncompressed before function evaluation). Because of this, all combinations of const and non-const arguments must generally be handled (except they are disallowed by other means but that is not the case here). So for 2 arguments, there can be
(1) const value, const timezone
(2) const value, non-const timezone
(3) non-const value, const timezone
(4) non-const value, non-const timezone.

Before this PR, the code only handled the non-const value cases (3) and (4). And the only test we had checked case (1) aka. const/const. Because of `useDefaultImplementationForConstants() == 1`, the arguments were materialized into case (4) and everything worked fine. So basically this was bad test coverage, but nobody noticed because the snowflake functions are obscure (imho).

The non-const timezone cases (2) and (4) are even more obscure since the time zone is part of the type, not of the value, and misleading, and because of that PR #48471 made them illegal in April. This broke old table definitions, so I made PR #50834 which adds a setting to switch back to the old behavior if needed but keeps non-const timezones illegal by default. AST Fuzzer in CI of #48471 then generated a query of type (2) combined with the old behavior as a setting. Because `useDefaultImplementationForConstants` only has an effect if *all* arguments are const, and because the code did not handle const values, this query went **boom**. So this PR finally allows case (2), even if it is super obscure. And again: Case (1) was transformed to (4), and cases (3) and (3) never were a problem.

Functions `DateTime*ToSnowflake` accept only a single argument, i.e. have cases (1) const value or (2) non-const value. Case (2) works fine and case (1) is materialized into case (2). So nothing to fix there.

Reply:
Thanks for the clear and detailed explanation! If we had a nomination for the clarity among comments in the repo, this one would be definitely a candidate for the nomination :)